src/mlmc/sampling_pool.py

```diff
@@ -112,53 +185,36 @@ def _queues_to_list(self, queue_dict):
 
         return results
 
-    # def _modify_level_sim_obj(self, level_sim, level_id, sample_dir):
-    #     """
-    #     Copy simulation common files to sample directory
-    #     :param level_sim: LevelSimulation instance
-    #     :param level_id: level identifier
-    #     :param sample_dir: sample directory
-    #     :return: LevelSimulation
-    #     """
-    #     # Copy simulation common files
-    #     if level_sim.common_files:
-    #         self.copy_sim_files(level_sim.common_files, sample_dir)
-    #
-    #         if level_sim.sample_workspace is not None:
-    #             raise Exception("Sample workspace must be set in Sampler method")
-    #         level_sim.sample_workspace = sample_dir
-    #
-    #     return level_sim
-    #
-    # def copy_sim_files(self, files: List[str], sample_dir):
-    #     """
-    #     Copy simulation common files to current simulation sample directory
-    #     :param files: List of files
-    #     :return:
-    #     """
-    #     for file in files:
-    #         shutil.copy(file, sample_dir)
-
-
-# class ProcessPool(OneProcessPool):
-#
-#     def __init__(self, n_processes):
-#         self._pool = ProcPool(n_processes)
-#         self._queues = {}
-#         self._n_running = 0
-#
-#         self._queue = queue.Queue()
-#
-#     def schedule_sample(self, sample_id, level_sim):
-#         level_sim.config_dict["sample_id"] = sample_id
-#         result = self._pool.apply_async(ProcessPool.calculate_sample, args=(sample_id, level_sim, ))
-#         res = result.get()
-#         self._queues.setdefault(level_sim.level_id, queue.Queue()).put((res[0], res[1][0], res[1][1]))
-#         return result
-
-#
-# class ThreadPool(ProcessPool):
-#
-#     def __init__(self, n_thread):
-#         self._pool = Threads(n_thread)
-#         self._queue = queue.Queue()
+
+class ProcessPool(OneProcessPool):
+
+    def __init__(self, n_processes, work_dir=None):
+        self._pool = ProcPool(n_processes)
+        super().__init__(work_dir=work_dir)
+
+    def schedule_sample(self, sample_id, level_sim):
+        self._n_running += 1
+
+        if level_sim.need_sample_workspace:
+            self._handle_sim_files(sample_id, level_sim)
+
+        res = self._pool.apply_async(SamplingPool.calculate_sample,
+                                                            args=(sample_id, level_sim,))
+
+        sample_id, result, err_msg = res.get()
+
+        if not err_msg:
+            self._queues.setdefault(level_sim.level_id, queue.Queue()).put((sample_id, (result[0], result[1])))
+        else:
+            self._failed_queues.setdefault(level_sim.level_id, queue.Queue()).put((sample_id, err_msg))
+            self._move_failed_dir(sample_id)
+
+
+class ThreadPool(ProcessPool):
```

Comment:
Add a comment about typical use cases:

MultiProcessPool - local parallel sampling for simulations WITHOUT external program call
ThreadPool - local parallel sampling for simulations WITH external program call